### PR TITLE
ARM: cortex-M: qemu: add support for executing tests with FPU/FPU_SHARING

### PR DESCRIPTION
**Pull request by ioannisg**
We are adding support for running tests and samples in CI for QEMU on Cortex-M with FPU and FPU_SHARING options enabled. **This fixes a severe limitation we had until now, by not covering at all the ARCH/Kernel implementation around Floating Point support in QEMU, at least not for Cortex-M. This is important for Zephyr LTS release.**

The solution is a bit complex, and involves extending the remote-core (CPU1) target of MPS2 AN521, since FPU support is only emulated on CPU1. The patch re-organizes a bit the AN521 SoC and board(s) definitions accordingly.

Effectively, we add support (locally for AN521) for multi-image builds, because the CPU1 binaries cannot run as standalone images. While it is desired to have a generic multi-image solution, we currently restricted the implementation to the AN521 board, to address our (important) use-case, without making drastic changes to Zephyr build framework.

### CODE REVIEW

File: boards/arm/mps2_an521/CMakeLists.txt
Comment (tejlmand) [inline boards/arm/mps2_an521/CMakeLists.txt:19]: why is this needed here ?

File: boards/arm/mps2_an521/CMakeLists.txt
Comment (tejlmand) [inline boards/arm/mps2_an521/CMakeLists.txt:25]: I don't find `empty_cpu0` anywhere ?

File: boards/arm/mps2_an521/CMakeLists.txt
Comment (tejlmand) [inline boards/arm/mps2_an521/CMakeLists.txt]: If user builds a dedicated sample for CPU0 himself, then how can he avoid building the `empty_cpu0` sample ?

File: boards/arm/mps2_an521/board.cmake
Comment (tejlmand) [inline boards/arm/mps2_an521/board.cmake:22]: This looks something that could easily break in future.

File: boards/arm/mps2_an521/CMakeLists.txt
Comment (tejlmand) [inline boards/arm/mps2_an521/CMakeLists.txt:11]: should this be more generic ?
Afterall, OpenAmp might not be the only sample that should disable the empty-cpu0.

**Review by tejlmand — COMMENTED** [review]
The initial approach looks sane.

A couple of small observations.

File: boards/arm/mps2_an521/CMakeLists.txt
Comment (ioannisg) [reply boards/arm/mps2_an521/CMakeLists.txt:11]: Well, it could. Although, only openAMP samples/applications/tests for Cortex-M are currently multi-core.
Honestly, I would not introduce a new abstract option, since we only use this, currently, in the MPS2 AN521 platform.

File: boards/arm/mps2_an521/board.cmake
Comment (ioannisg) [reply boards/arm/mps2_an521/board.cmake:22]: I leave this up to you, @tejlmand , to propose a better solution to this :)
For now I used what was already the solution in the rpmsg samples in the tree. 
Would welcome a proposal to improve this.

File: boards/arm/mps2_an521/CMakeLists.txt
Comment (ioannisg) [reply boards/arm/mps2_an521/CMakeLists.txt]: Looks, like we cannot get around without a new Kconfig option indicating dual core builds, e.g. config BOARD_MPS2_AN521_BUILD_FOR_STANDALONE_CPU1

File: boards/arm/mps2_an521/CMakeLists.txt
Comment (ioannisg) [reply boards/arm/mps2_an521/CMakeLists.txt:25]: thanks! fixed

File: boards/arm/mps2_an521/CMakeLists.txt
Comment (ioannisg) [reply boards/arm/mps2_an521/CMakeLists.txt:19]: possibly some redundant copy-paste
will check.

**Review by stephanosio — APPROVED** [review]
It looks ok overall.

Just some notes:
1. cmsis-dsp tests should be updated to test with `mps2_an521_remote` (w/ hardfp) after this is merged (it is currently being tested using `mps2_an521` with softfp).
2. `v2m_musca_b1` has FPU enabled on the CPU0, so we could consider using it as the default v8M testing platform in the future:
https://github.com/qemu/qemu/blob/a97978bcc2d1f650c7d411428806e5b03082b8c7/hw/arm/musca.c#L390-L397

File: boards/arm/mps2_an521/CMakeLists.txt
Comment (ioannisg) [reply boards/arm/mps2_an521/CMakeLists.txt]: Removed now, thanks!

File: boards/arm/mps2_an521/CMakeLists.txt
Comment (ioannisg) [reply boards/arm/mps2_an521/CMakeLists.txt:11]: @tejlmand, I thought about it and my feeling is that we are OK with the comment that is placed here, inline.
We could have a CONFIG_DUAL_CORE_APPLICATION option to instruct the build, here, _not_ to build an empty CPU0, but since openamp is the only in-tree subsystem that currently relates to multi-core applications, I would skip introducing it in favor of having the explanatory comment.

**Comment by ioannisg** [comment]
> It looks ok overall.
> 
> Just some notes:
> 
> 1. cmsis-dsp tests should be updated to test with `mps2_an521_remote` (w/ hardfp) after this is merged (it is currently being tested using `mps2_an521` with softfp).
> 2. `v2m_musca_b1` has FPU enabled on the CPU0, so we could consider using it as the default v8M testing platform in the future:
>    https://github.com/qemu/qemu/blob/a97978bcc2d1f650c7d411428806e5b03082b8c7/hw/arm/musca.c#L390-L397

Opened ticket: https://github.com/zephyrproject-rtos/zephyr/issues/36013

**Review by dcpleung — APPROVED** [review]

File: boards/arm/mps2_an521/Kconfig.board
Comment (microbuilder) [inline boards/arm/mps2_an521/Kconfig.board:10]: also a nit, but this feels a bit unnatural to me, and `BOARD_MPS2_AN521_CPU0_NS` feels easier to read, and is more consistent with current `_ns` or `_nonsecure` naming conventions?

File: boards/arm/mps2_an521/doc/index.rst
Comment (microbuilder) [inline boards/arm/mps2_an521/doc/index.rst:53]:
```suggestion
| mps2_an521_remote    | For building firmware on CPU1                        |
```

File: boards/arm/mps2_an521/doc/index.rst
Comment (microbuilder) [inline boards/arm/mps2_an521/doc/index.rst:365]: nit: Too much space

File: boards/arm/mps2_an521/doc/index.rst
Comment (microbuilder) [inline boards/arm/mps2_an521/doc/index.rst:431]:
```suggestion
Applications may be built for the second Cortex-M33
```

File: boards/arm/mps2_an521/CMakeLists.txt
Comment (microbuilder) [inline boards/arm/mps2_an521/CMakeLists.txt:15]:
```suggestion
# any binary for CPU0, as this is built by the dual core sample.
```

**Comment by microbuilder** [comment]
Others than some minor documentation cleanup, looks good to me, and nice to finally be able to emulate FP instructions in CI. Thanks for putting this together.

File: boards/arm/mps2_an521/Kconfig.board
Comment (ioannisg) [reply boards/arm/mps2_an521/Kconfig.board:10]: the Nordic boards have `ns` and `NS` without the underscores, so we lack consistency. I 'll change this, anyways, to what you suggest. But i think we need to agree on some rules here for the Non-Secure variant naming.

Comment (ioannisg) [reply boards/arm/mps2_an521/Kconfig.board:10]: Fixed.

**Comment by ioannisg** [comment]
> Others than some minor documentation cleanup, looks good to me, and nice to finally be able to emulate FP instructions in CI. Thanks for putting this together.

Thanks @microbuilder , comments addressed :)

File: boards/arm/mps2_an521/board.cmake
Comment (ioannisg) [reply boards/arm/mps2_an521/board.cmake:22]: @tejlmand any change you can provide a hint here?

**Review by microbuilder — APPROVED** [review]
Tested with `scripts/twister -p mps2_an521_remote -T tests/kernel`

**Comment by MartinSchoenstedt** [comment]
In the documentation changes you made you stated, that only CPU0 has the security extension enabled. However, for the AN521 actually both CPUs include TrustZone. I cannot find the documentation for this anymore but I have implemented a dual core system on the AN521 where both CPUs use the secure and non-secure hardware, configure their SAUs etc.

**Comment by koffes** [comment]
@ioannisg : Great addition! I tried using this, but we have a dependency on `FP_HARDABI`. Is this supported yet?

**Pull request merged**
